keras_cv/layers/object_detection/roi_align.py

```diff
@@ -0,0 +1,384 @@
+        2, output_size * 2, num_filters].
+      kernel_y: Tensor of size [batch_size, boxes, output_size, 2, 1].
+      kernel_x: Tensor of size [batch_size, boxes, output_size, 2, 1].
+
+    Returns:
+      A 5-D tensor representing feature crop of shape
+      [batch_size, num_boxes, output_size, output_size, num_filters].
+    """
+    features_shape = tf.shape(features)
+    batch_size, num_boxes, output_size, num_filters = (
+        features_shape[0],
+        features_shape[1],
+        features_shape[2],
+        features_shape[4],
+    )
+
+    output_size = output_size // 2
+    kernel_y = tf.reshape(kernel_y, [batch_size, num_boxes, output_size * 2, 1])
+    kernel_x = tf.reshape(kernel_x, [batch_size, num_boxes, 1, output_size * 2])
+    # Use implicit broadcast to generate the interpolation kernel. The
+    # multiplier `4` is for avg pooling.
+    interpolation_kernel = kernel_y * kernel_x * 4
+
+    # Interpolate the gathered features with computed interpolation kernels.
+    features *= tf.cast(
+        tf.expand_dims(interpolation_kernel, axis=-1), dtype=features.dtype
+    )
+    features = tf.reshape(
+        features,
+        [batch_size * num_boxes, output_size * 2, output_size * 2, num_filters],
+    )
+    features = tf.nn.avg_pool(features, [1, 2, 2, 1], [1, 2, 2, 1], "VALID")
+    features = tf.reshape(
+        features, [batch_size, num_boxes, output_size, output_size, num_filters]
+    )
+    return features
+
+
+def _compute_grid_positions(
+    boxes: tf.Tensor, boundaries: tf.Tensor, output_size: int, sample_offset: float
+) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor, tf.Tensor]:
+    """
+    Computes the grid position w.r.t. the corresponding feature map.
+
+    Args:
+      boxes: a 3-D tensor of shape [batch_size, num_boxes, 4] encoding the
+        information of each box w.r.t. the corresponding feature map.
+        boxes[:, :, 0:2] are the grid position in (y, x) (float) of the top-left
+        corner of each box. boxes[:, :, 2:4] are the box sizes in (h, w) (float)
+          in terms of the number of pixels of the corresponding feature map size.
+      boundaries: a 3-D tensor of shape [batch_size, num_boxes, 2] representing
+        the boundary (in (y, x)) of the corresponding feature map for each box.
+        Any resampled grid points that go beyond the bounary will be clipped.
+      output_size: a scalar indicating the output crop size.
+      sample_offset: a float number in [0, 1] indicates the subpixel sample offset
+        from grid point.
+
+    Returns:
+      kernel_y: Tensor of size [batch_size, boxes, output_size, 2, 1].
+      kernel_x: Tensor of size [batch_size, boxes, output_size, 2, 1].
+      box_grid_y0y1: Tensor of size [batch_size, boxes, output_size, 2]
+      box_grid_x0x1: Tensor of size [batch_size, boxes, output_size, 2]
+    """
+    boxes_shape = tf.shape(boxes)
+    batch_size, num_boxes = boxes_shape[0], boxes_shape[1]
+    if batch_size is None:
```

Comment:
I don't think tf.shape() can return None for any of the dimension. It is only possible for `boxes.shape`